cpp/oneapi/dal/backend/primitives/sort/sort_dpc.cpp

```diff
@@ -0,0 +1,534 @@
+
+    auto event = queue.submit([&](sycl::handler& cgh) {
+        cgh.depends_on(deps);
+        cgh.parallel_for(nd_range, [=](sycl::nd_item<1> item) {
+            auto sbg = item.get_sub_group();
+            if (sbg.get_group_id() > 0) {
+                return;
+            }
+
+            const std::uint32_t n_groups = item.get_group_range(0);
+            const std::uint32_t n_sub_groups = sbg.get_group_range()[0];
+            const std::uint32_t n_total_sub_groups = n_sub_groups * n_groups;
+            const Index elems_for_sbg =
+                elem_count / n_total_sub_groups + bool(elem_count % n_total_sub_groups);
+            const std::uint32_t local_size = sbg.get_local_range()[0];
+
+            const std::uint32_t local_id = sbg.get_local_id();
+            const std::uint32_t sub_group_id = sbg.get_group_id();
+            const std::uint32_t group_id = item.get_group().get_id(0) * n_sub_groups + sub_group_id;
+
+            Index ind_start = group_id * elems_for_sbg;
+            Index ind_end =
+                sycl::min(static_cast<Index>((group_id + 1) * elems_for_sbg), elem_count);
+
+            Index offset[radix_range_];
+
+            for (std::uint32_t i = 0; i < radix_range_; i++) {
+                offset[i] = part_prefix_hist_ptr[group_id * radix_range_ + i] +
+                            part_prefix_hist_ptr[n_total_sub_groups * radix_range_ + i];
+            }
+
+            for (Index i = ind_start + local_id; i < ind_end; i += local_size) {
+                radix_integer_t data_value = val_in_ptr[i];
+                radix_integer_t data_bits = ((inv_bits(data_value) >> bit_offset) & radix_range_1_);
+                Index pos_new = 0;
+                for (std::uint32_t j = 0; j < radix_range_; j++) {
+                    Index value = static_cast<Index>(data_bits == j);
+                    Index boundary = exclusive_scan(sbg, value, plus<Index>());
+                    pos_new |= value * (offset[j] + boundary);
+                    Index partial_offset = reduce(sbg, value, plus<Index>());
+                    offset[j] = offset[j] + partial_offset;
+                }
+                val_out_ptr[pos_new] = data_value;
+                ind_out_ptr[pos_new] = ind_in_ptr[i];
+            }
+        });
+    });
+
+    return event;
+}
+
+template <typename Float, typename Index>
+radix_sort_indices_inplace<Float, Index>::radix_sort_indices_inplace(const sycl::queue& queue)
+        : queue_(queue),
+          elem_count_(0) {}
+
+template <typename Float, typename Index>
+radix_sort_indices_inplace<Float, Index>::~radix_sort_indices_inplace() {
+    sort_event_.wait_and_throw();
+}
+
+template <typename Float, typename Index>
+void radix_sort_indices_inplace<Float, Index>::init(sycl::queue& queue, std::int64_t elem_count) {
+    ONEDAL_ASSERT(elem_count > 0);
+    if (elem_count > de::limits<Index>::max()) {
```

Comment:
Because if Index is int32 then elem_count can't exceed int32_max, due to there will be incorrect vals in index array if it exceeds it. And there is no problem with converting it into uint32 in this case.

Reply:
To make it simple return back std::uint32_t